no SIGSTKFLT on sparc?

Hi,

rust-1.64.0 added vendored `rustix-0.35.6`

and now rustc fails to build on sparc.


I think `sparc/sparc64` needs similar treatment to `mips/mips64` wrt `SIGSTKFLT`

downstream bug: https://bugs.gentoo.org/872725

full build log: https://bugs.gentoo.org/attachment.cgi?id=814030

build log snip:

```shell-session
error[E0425]: cannot find value `SIGSTKFLT` in crate `c`
   --> /var/tmp/portage/dev-lang/rust-1.64.0/work/rustc-1.64.0-src/vendor/rustix/src/imp/libc/process/types.rs:205:17
    |
205 |     Stkflt = c::SIGSTKFLT,
    |                 ^^^^^^^^^ help: a constant with a similar name exists: `SIGSTKSZ`
    |
   ::: /var/tmp/portage/dev-lang/rust-1.64.0/work/rustc-1.64.0-src/vendor/libc/src/unix/linux_like/linux/gnu/b64/sparc64/mod.rs:446:1
    |
446 | pub const SIGSTKSZ: ::size_t = 16384;
    | ------------------------------------- similarly named constant `SIGSTKSZ` defined here

error[E0531]: cannot find unit struct, unit variant or constant `SIGSTKFLT` in crate `c`
[...]
error: could not compile `rustix` due to 3 previous errors; 10 warnings emitted

Caused by:
  process didn't exit successfully: `CARGO=/usr/lib/rust/1.63.0/bin/cargo CARGO_CRATE_NAME=rustix CARGO_MANIFEST_DIR=/var/tmp/portage/dev-lang/rust-1.64.0/work/rustc-1.64.0-src/vendor/rustix CARGO_PKG_AUTHORS='Dan Gohman <[dev@sunfishcode.online](mailto:dev@sunfishcode.online)>:Jakub Konka <[kubkon@jakubkonka.com](mailto:kubkon@jakubkonka.com)>' CARGO_PKG_DESCRIPTION='Safe Rust bindings to POSIX/Unix/Linux/Winsock2-like syscalls' CARGO_PKG_HOMEPAGE='' CARGO_PKG_LICENSE='Apache-2.0 WITH LLVM-exception OR Apache-2.0 OR MIT' CARGO_PKG_LICENSE_FILE='' CARGO_PKG_NAME=rustix CARGO_PKG_REPOSITORY='https://github.com/bytecodealliance/rustix' CARGO_PKG_RUST_VERSION='' CARGO_PKG_VERSION=0.35.6 CARGO_PKG_VERSION_MAJOR=0 CARGO_PKG_VERSION_MINOR=35 CARGO_PKG_VERSION_PATCH=6 CARGO_PKG_VERSION_PRE='' LD_LIBRARY_PATH='/var/tmp/portage/dev-lang/rust-1.64.0/work/rustc-1.64.0-src/build/bootstrap/debug/deps:/usr/lib/rust/1.63.0/lib:' OUT_DIR=/var/tmp/portage/dev-lang/rust-1.64.0/work/rustc-1.64.0-src/build/bootstrap/debug/build/rustix-c545a2ac80fd0af8/out /usr/lib/rust/1.63.0/bin/rustc --crate-name rustix --edition=2018 /var/tmp/portage/dev-lang/rust-1.64.0/work/rustc-1.64.0-src/vendor/rustix/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no -C debuginfo=0 --cfg 'feature="default"' --cfg 'feature="fs"' --cfg 'feature="io-lifetimes"' --cfg 'feature="std"' -C metadata=45510bcf649eb595 -C extra-filename=-45510bcf649eb595 --out-dir /var/tmp/portage/dev-lang/rust-1.64.0/work/rustc-1.64.0-src/build/bootstrap/debug/deps -L dependency=/var/tmp/portage/dev-lang/rust-1.64.0/work/rustc-1.64.0-src/build/bootstrap/debug/deps --extern bitflags=/var/tmp/portage/dev-lang/rust-1.64.0/work/rustc-1.64.0-src/build/bootstrap/debug/deps/libbitflags-239db34e64ae628d.rmeta --extern libc_errno=/var/tmp/portage/dev-lang/rust-1.64.0/work/rustc-1.64.0-src/build/bootstrap/debug/deps/liberrno-5a163241a2553a6f.rmeta --extern io_lifetimes=/var/tmp/portage/dev-lang/rust-1.64.0/work/rustc-1.64.0-src/build/bootstrap/debug/deps/libio_lifetimes-38c82b3d4423d885.rmeta --extern libc=/var/tmp/portage/dev-lang/rust-1.64.0/work/rustc-1.64.0-src/build/bootstrap/debug/deps/liblibc-f41fd3c181356920.rmeta --cap-lints warn -C linker=sparc64-unknown-linux-gnu-gcc -Wrust_2018_idioms -Wunused_lifetimes -Wsemicolon_in_expressions_from_macros -Dwarnings --cfg libc` (exit status: 1)
  ```

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -71,6 +71,8 @@ jobs:
         armv5te-unknown-linux-gnueabi
         s390x-unknown-linux-gnu
         arm-linux-androideabi
+        sparc64-unknown-linux-gnu
+        sparcv9-sun-solaris
         aarch64-linux-android
         aarch64-apple-ios
     - name: Install cross-compilation tools
@@ -111,6 +113,8 @@ jobs:
     - run: cargo check --workspace --release -vv --target=armv5te-unknown-linux-gnueabi --features=all-apis
     - run: cargo check --workspace --release -vv --target=s390x-unknown-linux-gnu --features=all-apis
     - run: cargo check --workspace --release -vv --target=arm-linux-androideabi --features=all-apis
+    - run: cargo check --workspace --release -vv --target=sparc64-unknown-linux-gnu --features=all-apis
+    - run: cargo check --workspace --release -vv --target=sparcv9-sun-solaris --features=all-apis
     - run: cargo check --workspace --release -vv --target=aarch64-apple-ios --features=all-apis
     - run: cargo check --workspace --release -vv --target=aarch64-linux-android --features=all-apis
 
@@ -187,6 +191,7 @@ jobs:
     - run: cargo check -Z build-std --target x86_64-unknown-openbsd --all-targets --features=all-apis
     - run: cargo check -Z build-std --target mips64-openwrt-linux-musl --all-targets --features=all-apis
     - run: cargo check -Z build-std --target x86_64-unknown-dragonfly --all-targets --features=all-apis
+    - run: cargo check -Z build-std --target sparc-unknown-linux-gnu --all-targets --features=all-apis
     # Omit --all-targets on haiku because not all the tests build yet.
     - run: cargo check -Z build-std --target x86_64-unknown-haiku --features=all-apis
     # x86_64-uwp-windows-msvc isn't currently working.

--- a/src/backend/libc/process/types.rs
+++ b/src/backend/libc/process/types.rs
@@ -221,7 +221,12 @@ pub enum Signal {
         target_os = "solaris",
         all(
             any(target_os = "android", target_os = "linux"),
-            any(target_arch = "mips", target_arch = "mips64"),
+            any(
+                target_arch = "mips",
+                target_arch = "mips64",
+                target_arch = "sparc",
+                target_arch = "sparc64"
+            ),
         )
     )))]
     Stkflt = c::SIGSTKFLT,
@@ -304,7 +309,12 @@ impl Signal {
                 target_os = "solaris",
                 all(
                     any(target_os = "android", target_os = "linux"),
-                    any(target_arch = "mips", target_arch = "mips64"),
+                    any(
+                        target_arch = "mips",
+                        target_arch = "mips64",
+                        target_arch = "sparc",
+                        target_arch = "sparc64"
+                    ),
                 )
             )))]
             c::SIGSTKFLT => Some(Self::Stkflt),

--- a/src/backend/libc/termios/types.rs
+++ b/src/backend/libc/termios/types.rs
@@ -766,6 +766,8 @@ pub const B2000000: Speed = c::B2000000;
 
 /// `B2500000`
 #[cfg(not(any(
+    target_arch = "sparc",
+    target_arch = "sparc64",
     target_os = "dragonfly",
     target_os = "freebsd",
     target_os = "haiku",
@@ -779,6 +781,8 @@ pub const B2500000: Speed = c::B2500000;
 
 /// `B3000000`
 #[cfg(not(any(
+    target_arch = "sparc",
+    target_arch = "sparc64",
     target_os = "dragonfly",
     target_os = "freebsd",
     target_os = "haiku",
@@ -792,6 +796,8 @@ pub const B3000000: Speed = c::B3000000;
 
 /// `B3500000`
 #[cfg(not(any(
+    target_arch = "sparc",
+    target_arch = "sparc64",
     target_os = "dragonfly",
     target_os = "freebsd",
     target_os = "haiku",
@@ -805,6 +811,8 @@ pub const B3500000: Speed = c::B3500000;
 
 /// `B4000000`
 #[cfg(not(any(
+    target_arch = "sparc",
+    target_arch = "sparc64",
     target_os = "dragonfly",
     target_os = "freebsd",
     target_os = "haiku",

--- a/src/backend/linux_raw/termios/types.rs
+++ b/src/backend/linux_raw/termios/types.rs
@@ -338,15 +338,19 @@ pub const B1500000: Speed = linux_raw_sys::general::B1500000;
 pub const B2000000: Speed = linux_raw_sys::general::B2000000;
 
 /// `B2500000`
+#[cfg(not(any(target_arch = "sparc", target_arch = "sparc64",)))]
 pub const B2500000: Speed = linux_raw_sys::general::B2500000;
 
 /// `B3000000`
+#[cfg(not(any(target_arch = "sparc", target_arch = "sparc64",)))]
 pub const B3000000: Speed = linux_raw_sys::general::B3000000;
 
 /// `B3500000`
+#[cfg(not(any(target_arch = "sparc", target_arch = "sparc64",)))]
 pub const B3500000: Speed = linux_raw_sys::general::B3500000;
 
 /// `B4000000`
+#[cfg(not(any(target_arch = "sparc", target_arch = "sparc64",)))]
 pub const B4000000: Speed = linux_raw_sys::general::B4000000;
 
 /// `CSIZE`

--- a/src/termios/constants.rs
+++ b/src/termios/constants.rs
@@ -44,6 +44,8 @@ pub use backend::termios::types::B1500000;
 )))]
 pub use backend::termios::types::B2000000;
 #[cfg(not(any(
+    target_arch = "sparc",
+    target_arch = "sparc64",
     target_os = "dragonfly",
     target_os = "freebsd",
     target_os = "haiku",
@@ -55,6 +57,8 @@ pub use backend::termios::types::B2000000;
 )))]
 pub use backend::termios::types::B2500000;
 #[cfg(not(any(
+    target_arch = "sparc",
+    target_arch = "sparc64",
     target_os = "dragonfly",
     target_os = "freebsd",
     target_os = "haiku",
@@ -66,6 +70,8 @@ pub use backend::termios::types::B2500000;
 )))]
 pub use backend::termios::types::B3000000;
 #[cfg(not(any(
+    target_arch = "sparc",
+    target_arch = "sparc64",
     target_os = "dragonfly",
     target_os = "freebsd",
     target_os = "haiku",
@@ -77,6 +83,8 @@ pub use backend::termios::types::B3000000;
 )))]
 pub use backend::termios::types::B3500000;
 #[cfg(not(any(
+    target_arch = "sparc",
+    target_arch = "sparc64",
     target_os = "dragonfly",
     target_os = "freebsd",
     target_os = "haiku",
@@ -788,6 +796,8 @@ pub fn speed_value(speed: backend::termios::types::Speed) -> Option<u32> {
         )))]
         backend::termios::types::B2000000 => Some(2_000_000),
         #[cfg(not(any(
+            target_arch = "sparc",
+            target_arch = "sparc64",
             target_os = "dragonfly",
             target_os = "freebsd",
             target_os = "haiku",
@@ -799,6 +809,8 @@ pub fn speed_value(speed: backend::termios::types::Speed) -> Option<u32> {
         )))]
         backend::termios::types::B2500000 => Some(2_500_000),
         #[cfg(not(any(
+            target_arch = "sparc",
+            target_arch = "sparc64",
             target_os = "dragonfly",
             target_os = "freebsd",
             target_os = "haiku",
@@ -810,6 +822,8 @@ pub fn speed_value(speed: backend::termios::types::Speed) -> Option<u32> {
         )))]
         backend::termios::types::B3000000 => Some(3_000_000),
         #[cfg(not(any(
+            target_arch = "sparc",
+            target_arch = "sparc64",
             target_os = "dragonfly",
             target_os = "freebsd",
             target_os = "haiku",
@@ -821,6 +835,8 @@ pub fn speed_value(speed: backend::termios::types::Speed) -> Option<u32> {
         )))]
         backend::termios::types::B3500000 => Some(3_500_000),
         #[cfg(not(any(
+            target_arch = "sparc",
+            target_arch = "sparc64",
             target_os = "dragonfly",
             target_os = "freebsd",
             target_os = "haiku",

--- a/src/termios/mod.rs
+++ b/src/termios/mod.rs
@@ -52,6 +52,8 @@ pub use constants::B1500000;
 )))]
 pub use constants::B2000000;
 #[cfg(not(any(
+    target_arch = "sparc",
+    target_arch = "sparc64",
     target_os = "dragonfly",
     target_os = "freebsd",
     target_os = "haiku",
@@ -63,6 +65,8 @@ pub use constants::B2000000;
 )))]
 pub use constants::B2500000;
 #[cfg(not(any(
+    target_arch = "sparc",
+    target_arch = "sparc64",
     target_os = "dragonfly",
     target_os = "freebsd",
     target_os = "haiku",
@@ -74,6 +78,8 @@ pub use constants::B2500000;
 )))]
 pub use constants::B3000000;
 #[cfg(not(any(
+    target_arch = "sparc",
+    target_arch = "sparc64",
     target_os = "dragonfly",
     target_os = "freebsd",
     target_os = "haiku",
@@ -85,6 +91,8 @@ pub use constants::B3000000;
 )))]
 pub use constants::B3500000;
 #[cfg(not(any(
+    target_arch = "sparc",
+    target_arch = "sparc64",
     target_os = "dragonfly",
     target_os = "freebsd",
     target_os = "haiku",

--- a/tests/time/y2038.rs
+++ b/tests/time/y2038.rs
@@ -14,6 +14,7 @@
 #[cfg(not(all(target_env = "musl", target_pointer_width = "32")))]
 #[cfg(not(all(target_os = "android", target_pointer_width = "32")))]
 #[cfg(not(all(target_os = "emscripten", target_pointer_width = "32")))]
+#[cfg(not(all(target_os = "linux", target_arch = "sparc")))]
 #[test]
 fn test_y2038() {
     use rustix::time::{Secs, Timespec};

